File: CommonFramework/Assets/CScripts/Components/TimerManager.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System;


//最小精度是1帧。。
public class TimerManager : MonoBehaviour 
{
	class TimerEvent
	{
		public int id;
		public float delay;
		public Action<object> actionCallback;
		public object parm;
		public float timeCal;
		public int repeatTimes;
		public int currentTime;
	}

	private int m_id = 0;
	DictionaryNoLeak<int, TimerEvent> m_dicTimers = new DictionaryNoLeak<int, TimerEvent>();
	List<int> listIDWaitToRemove = new List<int>();

	private static TimerManager m_Instance;
	public static TimerManager Instance
	{
		get
		{
			return m_Instance;
		}
	}

	static bool initialized;

	public static void Initialize()
	{
		if (!initialized)
		{

			if (!Application.isPlaying)
				return;
			initialized = true;
			var g = new GameObject("TimerManager");
			DontDestroyOnLoad(g);
			m_Instance = g.AddComponent<TimerManager>();
		}

	}

	/// <summary>
	/// Calls the delay.
	/// </summary>
	/// <param name="actionDelay">Action delay.</param>
	/// <param name="delay">单位是秒</param>
	/// <param name="parm">Parm.</param>
	/// <param name="repeatTimes">重复次数，0为不重复，-1为无限重复,大于0为有限次</param>
	public int CallActionDelay(Action<object> actionDelay, float delay, object parm, int repeatTimes = 0)
	{
		if (delay == 0f)
		{
			delay = 1f;
		}
		if(repeatTimes < -1)
		{
			repeatTimes = -1;
		}
		TimerEvent timerEvent = new TimerEvent();
		timerEvent.id = m_id;

		timerEvent.actionCallback = actionDelay;
		timerEvent.delay = delay;
		timerEvent.parm = parm;
		timerEvent.repeatTimes = repeatTimes;
		m_dicTimers.AddKeyValue(m_id, timerEvent);

		m_id++;

		return timerEvent.id;
	}

	void DicTimersForEach(int id,TimerEvent timerEvent)
	{
		timerEvent.timeCal += Time.deltaTime;
		if(timerEvent.repeatTimes == 0)
		{
			if(timerEvent.timeCal >= timerEvent.delay)
			{
				if (timerEvent.actionCallback != null)
				{
					timerEvent.actionCallback(timerEvent.parm);
				}
				listIDWaitToRemove.Add(timerEvent.id);
			}

		}
		else if(timerEvent.repeatTimes == -1)
		{
			if (timerEvent.timeCal >= (timerEvent.currentTime + 1) * timerEvent.delay)
			{
				timerEvent.currentTime++;
				if (timerEvent.actionCallback != null)
				{
					timerEvent.actionCallback(timerEvent.parm);
				}
			}
		}
		else
		{
			if(timerEvent.timeCal >= (timerEvent.currentTime + 1)* timerEvent.delay)
			{
				timerEvent.currentTime++;
				if (timerEvent.actionCallback != null)
				{
					timerEvent.actionCallback(timerEvent.parm);
				}
				if(timerEvent.currentTime >= timerEvent.repeatTimes)
				{
					listIDWaitToRemove.Add(timerEvent.id);
				}
			}
		}
	}

	public void DeleteTimer(int id)
	{
		if(m_dicTimers.ContainsKey(id))
			m_dicTimers.RemoveKeyValue(id);
	}

	// Update is called once per frame
	void Update () 
	{
		if(m_dicTimers.Count > 0)
		{
			m_dicTimers.Foreach(DicTimersForEach);
			if(listIDWaitToRemove.Count > 0)
			{
				for (int i = 0; i < listIDWaitToRemove.Count; i++)
				{
					m_dicTimers.RemoveKeyValue(listIDWaitToRemove[i]);
				}
			}

		}
	}
}

File: CommonFramework/Assets/Lua/LuaScript/SystemModule/Timer/TimerManager.lua
---
--- Created by luzhuqiu.
--- DateTime: 2017/6/16 上午11:11
---
TimerEvent = class()
function TimerEvent:ctor()
    self.id = 0
    self.delay = 0
    self.actionCallback = nil
    self.parm = nil
    self.timeCal = 0
    self.repeatTimes = 0
    self.currentTime = 0
end

TimerManager = class()

function TimerManager:GetInstance()
    if self.m_Instance == nil then
        self.m_Instance = TimerManager.new()
    end
    return self.m_Instance
end

function TimerManager:ctor()
    self.id = 0
    self.dicTimers = {}
    self.listIDWaitToRemove = {}
    FixedUpdateBeat:Add(TimerManager.FixedUpdate,self)
end

---delay:单位是秒
---parm: actionDelay传入参数
---repeatTimes:重复次数，0 为不重复和 1 效果相同，-1为无限重复,大于0为有限次
function TimerManager:CallActionDelay(actionDelay,delay,parm,repeatTimes)
    if delay == 0 then
        delay = 1
    end

    if repeatTimes <= -1 then
        repeatTimes = -1
    end

    local timerEvent = TimerEvent.new()
    timerEvent.id = self.id
    timerEvent.actionCallback = actionDelay
    timerEvent.delay = delay
    timerEvent.parm = parm
    timerEvent.repeatTimes = repeatTimes
    self.dicTimers[timerEvent.id] = timerEvent

    self.id = self.id + 1

    return timerEvent.id
end

function TimerManager:DeleteTimer(id)
    self.dicTimers[id] = nil
end

function TimerManager.FixedUpdate(self)
    for k,timerEvent in pairs(self.dicTimers) do
        timerEvent.timeCal = timerEvent.timeCal + Time.fixedDeltaTime
        if timerEvent.repeatTimes == 0 then
            if timerEvent.timeCal >= timerEvent.delay then
                if timerEvent.actionCallback ~= nil then
                    timerEvent.actionCallback(timerEvent.parm)
                end
                table.insert(self.listIDWaitToRemove,timerEvent.id)
            end
        elseif timerEvent.repeatTimes == -1 then
            if timerEvent.timeCal >= (timerEvent.currentTime + 1) * timerEvent.delay then
                timerEvent.currentTime = timerEvent.currentTime + 1
                if timerEvent.actionCallback ~= nil then
                    timerEvent.actionCallback(timerEvent.parm)
                end
            end
        else
            if timerEvent.timeCal >= (timerEvent.currentTime + 1)* timerEvent.delay then
                timerEvent.currentTime = timerEvent.currentTime + 1
                if timerEvent.actionCallback ~= nil then
                    timerEvent.actionCallback(timerEvent.parm)
                end
                if timerEvent.currentTime >= timerEvent.repeatTimes then
                    table.insert(self.listIDWaitToRemove,timerEvent.id)
                end
            end
        end
    end

    for i = 1,#self.listIDWaitToRemove do
        self.dicTimers[self.listIDWaitToRemove[i]] = nil
    end

end

